### PR TITLE
[FEATURE][callouts] Add control over anchor point for callout on label

### PR DESCRIPTION
**Pull request by nyalldawson**
This gives users control over where a callout should join to the label
text. (Previously, you only had control over where the callout would
join to the corresponding feature geometry).

Choices include:
- Closest point (previous behavior)
- Label Centroid
- Fixed corners: Top left/top right/bottom left/bottom right/etc

Data defined control over the label anchor is also possible

Helps fix some callout ugliness!

Previous result (closest point mode):

image

With new "centroid" placement:

image

### CODE REVIEW

**Comment by nyalldawson** [comment]
@nirvn care to review this one?

**Comment by nirvn** [comment]
@nyalldawson , I'll do more than that. Not only am I going to review this, I'll add the settings is the label propreties map tool dialog too :)

**Comment by nyalldawson** [comment]
> I'll add the settings is the label propreties map tool dialog too :)

wheeeee :smiley:

File: src/ui/callouts/widget_simplelinecallout.ui [hunk 1]
```diff
@@ -203,39 +216,43 @@
...
        <property name="text">
-        <string>Anchor point</string>
+        <string>Label anchor point</string>
        </property>
       </widget>
      </item>
-     <item row="5" column="1" colspan="2">
-      <widget class="QComboBox" name="mAnchorPointComboBox"></widget>
+     <item row="6" column="1" colspan="2">
+      <widget class="QComboBox" name="mLabelAnchorPointComboBox"/>
```
Comment (nirvn) [inline src/ui/callouts/widget_simplelinecallout.ui]: Needs to be added in the <tabstops> section.

File: tests/src/core/testqgscallout.cpp [hunk 2]
```diff
@@ -939,6 +949,527 @@ void TestQgsCallout::calloutDataDefinedOffsetFromLabel()
...
+  callout->dataDefinedProperties().setProperty( QgsCallout::LabelAnchorPointPosition, QgsProperty::fromExpression( QStringLiteral( "'TL'" ) ) );
+  settings.setCallout( callout );
+
+  vl->setLabeling( new QgsVectorLayerSimpleLabeling( settings ) );  // TODO: this should not be necessary!
+  vl->setLabelsEnabled( true );
+
+  QgsDefaultLabelingEngine engine;
+  engine.setMapSettings( mapSettings );
+  engine.addProvider( new QgsVectorLayerLabelProvider( vl, QString(), true, &settings ) );
+  //engine.setFlags( QgsLabelingEngine::RenderOutlineLabels | QgsLabelingEngine::DrawLabelRectOnly );
```
Comment (nirvn) [inline tests/src/core/testqgscallout.cpp]: Those commented out lines are meant to stay for some practical debugging use or should those be removed?

Comment (nyalldawson) [reply tests/src/core/testqgscallout.cpp]: I honestly have no idea -- they came from tests in the label engine test suite which have been used as a template everywhere else. I don't know if it's still a valid comment or not, so I've just been leaving it in place...

Comment (nirvn) [reply tests/src/core/testqgscallout.cpp]: Ok, well +1 for consistency, let's leave those in :)

File: src/ui/callouts/widget_simplelinecallout.ui (hunk 1 above)
Comment (nyalldawson) [reply src/ui/callouts/widget_simplelinecallout.ui]: Fixed

**Review by nirvn — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks @nirvn!

**Pull request merged**